File: 2022/day02/php/solution.php
#!/usr/bin/env php
<?php

function parseInput(string $filename): array {
    $lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    $rounds = [];
    foreach ($lines as $line) {
        $parts = explode(' ', trim($line));
        if (count($parts) === 2) {
            $rounds[] = $parts;
        }
    }
    return $rounds;
}

function part1(array $rounds): int {
    // X=Rock, Y=Paper, Z=Scissors
    // Shape scores: Rock=1, Paper=2, Scissors=3
    $shapeScore = ['X' => 1, 'Y' => 2, 'Z' => 3];

    // Outcome: 0=loss, 3=draw, 6=win
    // A=Rock, B=Paper, C=Scissors
    $outcomes = [
        'A' => ['X' => 3, 'Y' => 6, 'Z' => 0],  // Rock vs X/Y/Z
        'B' => ['X' => 0, 'Y' => 3, 'Z' => 6],  // Paper vs X/Y/Z
        'C' => ['X' => 6, 'Y' => 0, 'Z' => 3],  // Scissors vs X/Y/Z
    ];

    $total = 0;
    foreach ($rounds as [$opp, $me]) {
        $total += $shapeScore[$me] + $outcomes[$opp][$me];
    }
    return $total;
}

function part2(array $rounds): int {
    // X=lose, Y=draw, Z=win
    // What shape to play given opponent and desired outcome
    // Returns the shape we play (1=Rock, 2=Paper, 3=Scissors)
    $choices = [
        'A' => ['X' => 3, 'Y' => 1, 'Z' => 2],  // Rock: lose->Scissors, draw->Rock, win->Paper
        'B' => ['X' => 1, 'Y' => 2, 'Z' => 3],  // Paper: lose->Rock, draw->Paper, win->Scissors
        'C' => ['X' => 2, 'Y' => 3, 'Z' => 1],  // Scissors: lose->Paper, draw->Scissors, win->Rock
    ];

    $outcomeScore = ['X' => 0, 'Y' => 3, 'Z' => 6];

    $total = 0;
    foreach ($rounds as [$opp, $outcome]) {
        $total += $choices[$opp][$outcome] + $outcomeScore[$outcome];
    }
    return $total;
}

function main(): void {
    $scriptDir = dirname(__FILE__);
    $inputFile = $scriptDir . '/../input.txt';

    $rounds = parseInput($inputFile);

    echo 'Part 1: ' . part1($rounds) . PHP_EOL;
    echo 'Part 2: ' . part2($rounds) . PHP_EOL;
}

main();

File: 2022/day02/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input(filename)
  File.readlines(filename).map(&:strip).reject(&:empty?).map(&:split)
end

def part1(rounds)
  # X=Rock, Y=Paper, Z=Scissors
  # Shape scores: Rock=1, Paper=2, Scissors=3
  shape_score = { 'X' => 1, 'Y' => 2, 'Z' => 3 }

  # Outcome: 0=loss, 3=draw, 6=win
  # A=Rock, B=Paper, C=Scissors
  outcomes = {
    ['A', 'X'] => 3,  # Rock vs Rock = draw
    ['A', 'Y'] => 6,  # Rock vs Paper = win
    ['A', 'Z'] => 0,  # Rock vs Scissors = loss
    ['B', 'X'] => 0,  # Paper vs Rock = loss
    ['B', 'Y'] => 3,  # Paper vs Paper = draw
    ['B', 'Z'] => 6,  # Paper vs Scissors = win
    ['C', 'X'] => 6,  # Scissors vs Rock = win
    ['C', 'Y'] => 0,  # Scissors vs Paper = loss
    ['C', 'Z'] => 3   # Scissors vs Scissors = draw
  }

  rounds.sum { |opp, me| shape_score[me] + outcomes[[opp, me]] }
end

def part2(rounds)
  # X=lose, Y=draw, Z=win
  # What shape to play given opponent and desired outcome
  # Returns the shape score (1=Rock, 2=Paper, 3=Scissors)
  choices = {
    ['A', 'X'] => 3,  # Rock, need to lose -> Scissors
    ['A', 'Y'] => 1,  # Rock, need to draw -> Rock
    ['A', 'Z'] => 2,  # Rock, need to win -> Paper
    ['B', 'X'] => 1,  # Paper, need to lose -> Rock
    ['B', 'Y'] => 2,  # Paper, need to draw -> Paper
    ['B', 'Z'] => 3,  # Paper, need to win -> Scissors
    ['C', 'X'] => 2,  # Scissors, need to lose -> Paper
    ['C', 'Y'] => 3,  # Scissors, need to draw -> Scissors
    ['C', 'Z'] => 1   # Scissors, need to win -> Rock
  }

  outcome_score = { 'X' => 0, 'Y' => 3, 'Z' => 6 }

  rounds.sum { |opp, outcome| choices[[opp, outcome]] + outcome_score[outcome] }
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  rounds = parse_input(input_file)

  puts "Part 1: #{part1(rounds)}"
  puts "Part 2: #{part2(rounds)}"
end

main
